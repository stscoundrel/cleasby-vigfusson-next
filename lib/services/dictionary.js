const { getDictionary } = require('cleasby-vigfusson-dictionary');
const slugify = require('slugify');

const addSlugs = (words) => {
  const existingSlugs = {};

  const formattedWords = words.map((word) => {
    let slug = slugify(word.word).toLowerCase();

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1;
      slug = `${slug}-${existingSlugs[slug]}`;
    } else {
      existingSlugs[slug] = 1;
    }

    return {
      ...word,
      slug,
    };
  });

  return formattedWords;
};

export const getAllWords = () => {
  const words = getDictionary();

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words);

  return formattedWords;
};

export default {
  getAllWords,
};
