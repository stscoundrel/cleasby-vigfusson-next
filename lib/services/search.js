export const searchDictionary = (search, dictionary) => {
  const results = dictionary.filter((entry) => {
    let matchesSearch = false

    if (entry.word.includes(search)) {
      matchesSearch = true
    }

    if (entry.slug.includes(search)) {
      matchesSearch = true
    }

    entry.definitions.forEach((definition) => {
      if (definition.includes(search)) {
        matchesSearch = true
      }
    })

    return matchesSearch
  })

  return results
}

export default {
  searchDictionary,
}
