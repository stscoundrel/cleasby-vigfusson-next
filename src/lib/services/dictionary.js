const { getDictionary } = require('cleasby-vigfusson-dictionary')
const { VALID_AS_FIRST } = require('old-norse-alphabet')
const { slugifyWord, slugifyLetter } = require('../utils/slugs')

let cachedDictionary = null

const addSlugs = (words) => {
  const existingSlugs = {}

  const formattedWords = words.map((word) => {
    let slug = slugifyWord(word.word).toLowerCase()

    if (slug.charAt(0) === '-') {
      slug = slug.substring(1)
    }

    if (existingSlugs[slug]) {
      // Double slug, make unique.
      existingSlugs[slug] += 1
      slug = `${slug}-${existingSlugs[slug]}`
    } else {
      existingSlugs[slug] = 1
    }

    return {
      ...word,
      slug,
    }
  })

  return formattedWords
}

export const getAllWords = () => {
  if (cachedDictionary) return cachedDictionary

  const words = getDictionary()

  /**
   * Add URL safe slugs.
   */
  const formattedWords = addSlugs(words)

  cachedDictionary = formattedWords

  return formattedWords
}

export const getByLetter = (letter) => {
  const words = getAllWords()
  const byLetter = words.filter((entry) => (
    entry.word.charAt(0).toLowerCase() === letter.toLowerCase()))

  return byLetter
}

export const getWord = (word) => getAllWords().filter((entry) => entry.slug === word)[0]

export const getAlphabet = () => {
  const letters = [...VALID_AS_FIRST.filter((letter) => letter !== 'ǫ' && letter !== 'ø'), 'ö']

  const formattedLetters = letters.map((letter) => ({
    letter,
    slug: slugifyLetter(letter),
  }))

  return formattedLetters
}

export default {
  getAllWords,
  getByLetter,
  getWord,
  getAlphabet,
}
