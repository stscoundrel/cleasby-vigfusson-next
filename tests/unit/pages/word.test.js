import ReactDOM from 'react-dom/client'
import Word, { getStaticProps, getStaticPaths } from 'pages/word/[word]'
import renderer from 'react-test-renderer'
import { getAlphabet } from 'lib/services/dictionary'

const mockHandler = jest.fn()

/**
 * Mock router
 */
jest.mock('next/router', () => ({
  useRouter() {
    return {
      locale: undefined,
      defaultLocale: undefined,
      asPath: '/test',
      back: mockHandler,
    }
  },
}))

describe('Word page: render & usage', () => {
  const word = {
    word: 'Völva',
    definitions: [
      'u, f., also spelt völfa, gen. völu, pl. völur; völfu. or also völfur or voluur; gen. pl. does not occur; the nom. Vala is erroneous: [the etymology as well as the origin of this word is uncertain; but may not the Norse Völva and the Gr. σίβυλλα be relations? the identity in sense at least is very striking; the Gr. word first occurs in Aristoph., and then in Plato; may it not have been adopted from some Scythian tribe, for a word like this, if Greek, could hardly fail to occur in Homer? in völva an initial s, we suppose, has been lost (qs. svölva); in the Greek the ĭ would be an inserted vowel]:—a prophetess, sibyl, wise woman; völva, seiðkona, spákona (qq. v.) are synonymous. The ancient Sagas contain many remarkable records of the heathen wise-women or sibyls, who were held in honour and reverence; at the great feasts and sacrifices in the autumn, the völva (often a woman of rank) went with her troop of maidens through the country, where she, so to say, crowned the feast; she was seated on a high seat (seiðhjallr) in the hall, where she wrought her spells and sang her ‘weird-songs’ (varðlokur), after which the guests went past her one by one, and she told each his fate, or whatever else one wanted to know, e. g. the course of the coming winter and the like. The former part of the Völuspá is evidently conceived as the inspired song of a völva, seated on her high seat, and addressing Odin, while the gods listen to her words; and the latter part of the poem appears to be a kind of necromancy, or the raising of a dead völva, as also is the lay Vegtamskviða; sú kona var þar í bygð er Þorbjörg hét, ok var kölluð lítil völva, hón hafði áttar sér níu systr, ok vóru allar spákonur, en hón var ein þá á lífi, þat var háttr Þorbjargar um vetrum, at hón fór á veizlur ok buðu þeir menn henni mest heim er forvitni var á at vita forlög sín eða árferð, Þorf. Karl. Names of such wise women, Gróa völva, Edda; Heimlaug völva, Gullþ.; Heiðr völva, Landn. 173; Huldr völva, Yngl. S., Hkr. i. 21; to which add the ‘Weleda’ of Tacitus: class. passages are Þorf. S. ch. 3 (exceedingly interesting), Örvar-Odds S. ch. 3 (Fas. ii. 506), Vd. ch. 10: völu vél-spá, Vsp.; eru völfur allar frá Víðólfi, Hdl.; völva ok vís kona. Vtkv.; þá kom til völva sú er Gróa hét, Edda 58; var á því landi spákona sú er sagði fyrir örlög manna, … þeir fara til móts við völunna, Fb. ii. 28; ek fór í skóg til þín í völvu líki, Fas. i. 135; þóttusk menn vita at þar mundi verst hafa völu-leiði, Ld. 328; þá reið Óðinn fyrir austan dyrr þar er hann vissi völu leiði, Vtkv.; úrsvöl Gýmis völva, of Rán the goddess, Edda (in a verse); at hás völva valdi því bölvi, Kormak; in a bad sense, völva and skollvis kona, Hkv. 1. 34; Tacitus (Germ. ch. 8, 46, and Hist. iv. 61, 65, v. 22, 24) speaks of these practices, as also does Plutarch, Caesar ch. 19,— τα μαντεύματα των ἱερων γυναικων.',
      'COMPDS: Völuspá, Völvustaðir.',
    ],
    slug: 'volva',
  }

  const abbreviations = {
    common: [
      {
        abbreviation: 'ch.',
        explanation: 'chapter.',
      },
      {
        abbreviation: 'f.',
        explanation: 'feminine.',
      },
    ],
    works: [],
  }

  test('Does not crash', () => {
    const div = document.createElement('div')
    const root = ReactDOM.createRoot(div)
    root.render(
      <Word entry={word} letters={getAlphabet()} abbreviations={abbreviations} />,
    )
  })

  test('Matches snapshot', () => {
    const tree = renderer.create(
      <Word entry={word} letters={getAlphabet()} abbreviations={abbreviations} />,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  test('Returns null if entry is unavailable', () => {
    const tree = renderer.create(
      <Word entry={null} letters={getAlphabet()} abbreviations={abbreviations} />,
    ).toJSON()
    expect(tree).toBeNull()
  })

  test('Back button works', async () => {
    const tree = renderer.create(
      <Word entry={word} letters={getAlphabet()} abbreviations={abbreviations} />,
    )

    // Click back btn.
    await renderer.act(async () => {
      expect(mockHandler).not.toHaveBeenCalled()
      await tree.root.findByProps({ text: 'Back' }).props.action()

      // Assert mockrouter received a push.
      expect(mockHandler).toHaveBeenCalled()
      expect(mockHandler.mock.calls.length).toBe(1);
    })
  })
})

describe('Word page: data fetching', () => {
  test('getStaticPaths works', async () => {
    const expected = {
      paths: [],
      fallback: 'blocking',
    }

    const result = await getStaticPaths()

    expect(result).toMatchObject(expected)
  })

  test('getStaticProps works', async () => {
    const expected = {
      props: {
        entry: {
          word: 'stækja',
          definitions: ['u, f. <i>a bad stench.</i>'],
          slug: 'staekja',
        },
        abbreviations: {
          common: [
            {
              abbreviation: 'ch.',
              explanation: 'chapter.',
            },
            {
              abbreviation: 'f.',
              explanation: 'feminine.',
            },
          ],
          works: [],
        },
        letters: getAlphabet(),
      },
    }

    const result = await getStaticProps({ params: { word: 'staekja' } })

    expect(result).toEqual(expected)
  })

  test('getStaticProps returns 404 redirect for unkown words', async () => {
    const expected = {
      props: {},
      notFound: true,
    }

    const result = await getStaticProps({ params: { word: 'loremipsum' } })

    expect(result).toEqual(expected)
  })
})
