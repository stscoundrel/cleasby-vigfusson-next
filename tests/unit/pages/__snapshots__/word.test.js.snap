// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Word page: render & usage Matches snapshot 1`] = `
[
  <header>
    <div
      className="topBarSpacer"
    />
    <nav
      className="section"
    >
      <div
        className="topbar container"
      >
        <img
          alt="To home"
          height="30"
          loading="lazy"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
          src="/favicon-48x48.png"
          width="30"
        />
        <div
          aria-label="Open menu"
          className="hamburger"
          onClick={[Function]}
          role="button"
        >
          <span
            className="part"
          />
          <span
            className="part"
          />
          <span
            className="part"
          />
        </div>
      </div>
      <div
        className="content null container"
      >
        <ul
          className="list"
        >
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              a
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/a2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              á
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/b"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              b
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/d"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              d
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              e
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/e2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              é
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/f"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              f
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/g"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              g
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/h"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              h
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              i
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/i2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              í
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/j"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              j
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/k"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              k
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/l"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              l
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/m"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              m
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/n"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              n
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              o
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/o2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ó
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/p"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              p
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/r"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              r
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/s"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              s
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/t"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              t
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              u
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/u2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ú
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/v"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              v
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              y
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/y2"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ý
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/th"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              þ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/ae"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              æ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe3"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              œ
            </a>
          </li>
          <li
            className="listItem"
          >
            <a
              className="link"
              href="/letter/oe"
              onClick={[Function]}
              onMouseEnter={[Function]}
              onTouchStart={[Function]}
            >
              ö
            </a>
          </li>
        </ul>
        <form
          className="form"
          onSubmit={[Function]}
        >
          <input
            aria-label="Search"
            className="input"
            name="search"
            onChange={[Function]}
            type="search"
            value=""
          />
          <button
            className="button button--secondary button--no-margs"
            name="submit"
            type="submit"
          >
            Search
          </button>
        </form>
      </div>
    </nav>
    <nav
      className="section"
    >
      <div
        className="container"
      >
        <a
          className="link"
          href="/"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Cleasby & Vigfusson Dictionary
        </a>
        <a
          className="link"
          href="/letter/v"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Letter V
        </a>
        <a
          className="link"
          href="/word/volva"
          onClick={[Function]}
          onMouseEnter={[Function]}
          onTouchStart={[Function]}
        >
          Völva
        </a>
        <script
          dangerouslySetInnerHTML={
            {
              "__html": "{"@context":"https://schema.org/","@type":"BreadcrumbList","itemListElement":[{"@type":"ListItem","position":1,"name":"Cleasby & Vigfusson Dictionary","item":"undefined/"},{"@type":"ListItem","position":2,"name":"Letter V","item":"undefined/letter/v"},{"@type":"ListItem","position":3,"name":"Völva","item":"undefined/word/volva"}]}",
            }
          }
          type="application/ld+json"
        />
      </div>
    </nav>
  </header>,
  <main
    className="container"
  >
    <article
      className="section"
    >
      <header>
        <h1
          lang="non"
        >
          Völva
        </h1>
        <small
          className="subHeading"
        >
          Old Norse Dictionary - 
          völva
        </small>
        <p>
          Meaning of Old Norse word 
          <em>
            "
            Völva
            "
          </em>
           in English.
        </p>
      </header>
      <p>
        As defined by the Cleasby & Vigfusson Old Norse to English dictionary:
      </p>
      <p>
        <dfn
          className="capitalize"
        >
          Völva
        </dfn>
         Old Norse word can mean:
      </p>
      <dl
        className="definitionList"
      >
        <dt>
          <strong>
            Völva
          </strong>
        </dt>
        <dd
          className="itemDescription"
          dangerouslySetInnerHTML={
            {
              "__html": "u, <abbr title="feminine.">f.</abbr>, also spelt völfa, gen. völu, pl. völur; völfu. or also völfur or voluur; gen. pl. does not occur; the nom. Vala is erroneous: [the etymology as well as the origin of this word is uncertain; but may not the Norse Völva and the Gr. σίβυλλα be relations? the identity in sense at least is very striking; the Gr. word first occurs in Aristoph., and then in Plato; may it not have been adopted from some Scythian tribe, for a word like this, if Greek, could hardly fail to occur in Homer? in völva an initial s, we suppose, has been lost (qs. svölva); in the Greek the ĭ would be an inserted vowel]:—a prophetess, sibyl, wise woman; völva, seiðkona, spákona (qq. v.) are synonymous. The ancient Sagas contain many remarkable records of the heathen wise-women or sibyls, who were held in honour and reverence; at the great feasts and sacrifices in the autumn, the völva (often a woman of rank) went with her troop of maidens through the country, where she, so to say, crowned the feast; she was seated on a high seat (seiðhjallr) in the hall, where she wrought her spells and sang her ‘weird-songs’ (varðlokur), after which the guests went past her one by one, and she told each his fate, or whatever else one wanted to know, e. g. the course of the coming winter and the like. The former part of the Völuspá is evidently conceived as the inspired song of a völva, seated on her high seat, and addressing Odin, while the gods listen to her words; and the latter part of the poem appears to be a kind of necromancy, or the raising of a dead völva, as also is the lay Vegtamskviða; sú kona var þar í bygð er Þorbjörg hét, ok var kölluð lítil völva, hón hafði áttar sér níu systr, ok vóru allar spákonur, en hón var ein þá á lífi, þat var háttr Þorbjargar um vetrum, at hón fór á veizlur ok buðu þeir menn henni mest heim er forvitni var á at vita forlög sín eða árferð, Þor<abbr title="feminine.">f.</abbr> Karl. Names of such wise women, Gróa völva, Edda; Heimlaug völva, Gullþ.; Heiðr völva, Landn. 173; Huldr völva, Yngl. S., Hkr. i. 21; to which add the ‘Weleda’ of Tacitus: class. passages are Þor<abbr title="feminine.">f.</abbr> S. <abbr title="chapter.">ch.</abbr> 3 (exceedingly interesting), Örvar-Odds S. <abbr title="chapter.">ch.</abbr> 3 (Fas. ii. 506), Vd. <abbr title="chapter.">ch.</abbr> 10: völu vél-spá, Vsp.; eru völfur allar frá Víðólfi, Hdl.; völva ok vís kona. Vtkv.; þá kom til völva sú er Gróa hét, Edda 58; var á því landi spákona sú er sagði fyrir örlög manna, … þeir fara til móts við völunna, Fb. ii. 28; ek fór í skóg til þín í völvu líki, Fas. i. 135; þóttusk menn vita at þar mundi verst hafa völu-leiði, Ld. 328; þá reið Óðinn fyrir austan dyrr þar er hann vissi völu leiði, Vtkv.; úrsvöl Gýmis völva, of Rán the goddess, Edda (in a verse); at hás völva valdi því bölvi, Kormak; in a bad sense, völva and skollvis kona, Hkv. 1. 34; Tacitus (Germ. <abbr title="chapter.">ch.</abbr> 8, 46, and Hist. iv. 61, 65, v. 22, 24) speaks of these practices, as also does Plutarch, Caesar <abbr title="chapter.">ch.</abbr> 19,— τα μαντεύματα των ἱερων γυναικων.",
            }
          }
        />
      </dl>
      <dl
        className="definitionList"
      >
        <dt>
          <strong>
            Völva
          </strong>
        </dt>
        <dd
          className="itemDescription"
          dangerouslySetInnerHTML={
            {
              "__html": "COMPDS: Völuspá, Völvustaðir.",
            }
          }
        />
      </dl>
      <p>
        Possible runic inscription in 
        <em>
          Younger Futhark:
        </em>
        <span
          className="rune"
        >
          ᚢᚢᛚᚢᛅ
        </span>
      </p>
      <div
        className="abbreviations"
      >
        <div
          className="column"
        >
          <h4>
            Abbreviations used:
          </h4>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                ch.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              chapter.
            </dd>
          </dl>
          <dl
            className="wrap"
          >
            <dt
              className="abbreviation"
            >
              <strong>
                f.
              </strong>
            </dt>
            <dd
              className="explanation"
            >
              feminine.
            </dd>
          </dl>
        </div>
      </div>
    </article>
    <a
      className="button"
      onClick={[Function]}
    >
      Back
    </a>
    <div
      aria-label="Back to top"
      className="backToTop"
      onClick={[Function]}
      role="button"
    >
      ↑
    </div>
  </main>,
  <footer
    className="section"
  >
    <div
      className="container"
    >
      <section
        className="section"
      >
        <h2>
          About
        </h2>
        <p>
          Based on Cleasby & Vigfusson Dictionary Old Norse dictionary.
        </p>
        <p>
          <em>
            Icelandic-English
          </em>
           dictionary was started by Richard Cleasby and finished by Gudbrand Vigfusson.
        </p>
        <p>
          It was published in 1874, which leads to there being many public domain versions of the book available.
        </p>
      </section>
      <section
        className="section"
      >
        <h3>
          Old Norse language
        </h3>
        <p>
          Old Norse was a North Germanic language that was spoken by inhabitants of Scandinavia and their overseas settlements from about the 7th to the 15th centuries.
        </p>
        <p>
          Also known as "the viking language", "Old Nordic", or "Old Scandinavian"
        </p>
      </section>
      <div
        className="navs"
      >
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Dictionary project
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/cleasby-vigfusson-next"
                rel="noopener noreferrer"
                target="_blank"
              >
                Source code
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/cleasby-vigfusson-dictionary"
                rel="noopener noreferrer"
                target="_blank"
              >
                Data source
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/cleasby-vigfusson-abbreviations"
                rel="noopener noreferrer"
                target="_blank"
              >
                Abbreviations
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Related packages
          </h4>
          <ul>
            <li>
              <a
                href="https://github.com/stscoundrel/younger-futhark"
                rel="noopener noreferrer"
                target="_blank"
              >
                Younger Futhark Runes
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norse-alphabet"
                rel="noopener noreferrer"
                target="_blank"
              >
                Old Norse Alphabet
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/old-norse-alphabet-sort"
                rel="noopener noreferrer"
                target="_blank"
              >
                Old Norse Alphabet Sort
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/gangleri"
                rel="noopener noreferrer"
                target="_blank"
              >
                Gangleri
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/teljari"
                rel="noopener noreferrer"
                target="_blank"
              >
                Teljari
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/markari"
                rel="noopener noreferrer"
                target="_blank"
              >
                Markari
              </a>
            </li>
            <li>
              <a
                href="https://github.com/stscoundrel/spyrjari"
                rel="noopener noreferrer"
                target="_blank"
              >
                Spyrjari
              </a>
            </li>
          </ul>
        </nav>
        <nav
          className="nav"
        >
          <h4
            className="navTitle"
          >
            Quick links
          </h4>
          <ul
            className="navColumns"
          >
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                a
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/a2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                á
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/b"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                b
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/d"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                d
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                e
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/e2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                é
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/f"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                f
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/g"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                g
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/h"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                h
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                i
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/i2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                í
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/j"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                j
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/k"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                k
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/l"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                l
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/m"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                m
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/n"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                n
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                o
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/o2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ó
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/p"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                p
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/r"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                r
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/s"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                s
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/t"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                t
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                u
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/u2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ú
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/v"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                v
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                y
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/y2"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ý
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/th"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                þ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/ae"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                æ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe3"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                œ
              </a>
            </li>
            <li
              className="navColumnItem"
            >
              <a
                className="link"
                href="/letter/oe"
                onClick={[Function]}
                onMouseEnter={[Function]}
                onTouchStart={[Function]}
              >
                ö
              </a>
            </li>
          </ul>
        </nav>
      </div>
      <small
        className="copyright"
      >
        Copyright © 2020 - 2020
        <br />
        <a
          href="https://github.com/stscoundrel"
          rel="noopener noreferrer"
          target="_blank"
        >
          Sampo Silvennoinen / StScoundrel
        </a>
      </small>
    </div>
  </footer>,
]
`;
